mountpoint-s3/src/inode.rs

```diff
@@ -354,6 +358,79 @@ impl Superblock {
 
         Ok(LookedUp { inode, stat })
     }
+
+    /// Remove local-only empty directory, i.e., the ones created by mkdir.
+    /// It does not affects empty directories represented remotely with directory markers.  
+    pub async fn rmdir<OC: ObjectClient>(
+        &self,
+        client: &OC,
+        parent_ino: InodeNo,
+        name: &OsStr,
+    ) -> Result<(), InodeError> {
+        let LookedUp { inode, .. } = self.lookup(client, parent_ino, name).await?;
+
+        if inode.kind() == InodeKind::File {
+            return Err(InodeError::NotADirectory(inode.ino()));
+        }
+
+        let parent = self.inner.get(parent_ino)?;
+        let mut parent_state = parent.inner.sync.write().unwrap();
+
+        let mut inode_state = inode.inner.sync.write().unwrap();
+
+        match inode_state.write_status {
+            WriteStatus::LocalOpen => unreachable!("A directory cannot be in Local open state"),
+
+            WriteStatus::Remote => {
+                error!(parent = parent_ino, ?name, "rmdir called on a remote directory",);
+                return Err(InodeError::CannotRemoveRemoteDirectory(inode.ino()));
+            }
+
+            WriteStatus::LocalUnopened => {
+                let inode_kind_data = &inode_state.kind_data;
+                // A local directory can only contain local children.
+                match inode_kind_data {
+                    InodeKindData::File {} => unreachable!("Already checked that inode is a directory"),
+                    InodeKindData::Directory {
+                        writing_children,
+                        deleted,
+                        ..
+                    } => {
+                        if *deleted {
+                            return Err(InodeError::InodeDoesNotExist(inode.ino()));
+                        }
```

Comment:
@jamesbornholt was also discussing on it. When we mark an Inode as deleted, it essentially meant to be removed and still having it in the HashMap maybe considered as a bug. I am not very sure, which option to take.

Reply:
For me, it was weird - we are saying Inode doesn't exist and passing in the Inode that doesn't exist.

The distinct error would mean that the error cause is clear from reading (we found a directory now marked as deleted), and in any error cause output it would also be clear.

Reply:
Agree that keeping it around in the `HashMap` feels like a bug anyway. We aren't checking everywhere if the `Inode` was deleted.

Maybe we should be asserting/checking that the `Inode` does not have this flag set as part of `superblock.get(ino)`, but that involves another lock... :(

Reply:
I see. But aren't all these implementation details? A user of `Superblock` will call `rmdir` (or `lookup`) and get `InodeError::InodeDoesNotExist`. What would it mean to them if they got `InodeError::InodeMarkedAsDeleted` sometimes? Would they get any access to such a node?

Reply:
Yeah my point was that the distinction between `deleted` and just removing it from the hashmap is an internal implementation detail, not worth surfacing as a separate error since the caller both doesn't know about the detail and can't do anything with it.

Reply:
I'm OK with it being left as is then.